man/land.1
.TH LAND 1 "7 Aug 1995" "land"
.UC
.SH NAME
land \- Empire land generator
.SH SYNOPSIS
.B land
[
.IR sectrange
.IR mincontdist
.IR chunksize
.IR numchunks
.IR numisle
]
.br
.SH DESCRIPTION
While most deities prefer to use
.I fairland
to generate their land, this program is included as an alternative for
creating the geography for a game.
.SH OPTIONS
.TP
.IR sectrange
smoothing area (default = 3)
.TP
.IR mincontdist
minimum distance between continents (default = 2)
.TP
.IR chunksize
basic land block size (default = 2)
.TP
.IR numchunks
number of chunks per continent (default = 70)
.TP
.IR numisle
number of islands (default is based on world size)


doc/CHANGES
This file contains a list of patches/upgrades made to the server for a specific
version.  It also contains the name of the author of that particular code.  It
may not be complete.

*version 2.0 beta 1 release 1*
Fixed age.c so reserves degrade normally instead of growing! - Chad Zabel
Fixed arch.c - replaced wrapp with what was supposed to be there - Ken Huisman
Patched bug in atta.c - a temporary patch that could be done better
( and will be later ) - Ken H.
patched bug in lib/commands/mail.c - Dave Pare
added patches to make code compile under linux - Ken H. based on linux chain 3
patches by someone else whose name I forget.  Note at this time the code
*compiles* under linux, but there are still bugs - the server seems to
stop taking socket connections or something after someone logs out of the
game.
fixed a null pointer bug in lib/subs/bigmap.c - Chad Zabel

*version 2.0 beta 1 release 2*
Patch for exec bug - Chad Zabel

*version 2.0 beta 1 release 3*
Patch for zepplin bug - Jim Ortlieb
Added unixware(svr4.2) specific stuff to arch.c - Doug Hay
Changes to h/lwp.h and lib/lwp/lwp.c for thread stack checker - Doug Hay

*version 2.0 beta 1 release 4*
Modified fairland.c to 1.6.2 and made it so translate_continents is only
called when BLITZ is defined  - Ken Huisman, Ken Stevens
Modifications to lwp.h and arch.c to fix linux threads - Scott Ferguson

*version 2.0 beta 1 release 5*
Changed version number to 2.0.0 - Ken Huisman
fixed unterminated path bug in lib/common/path.c - Scott Ferguson
fixed first part of bug described in wish list item c6 ( false resupply ) -
Scott Ferguson
fixed part of bug described in wish list item t3 ( you have 666 new telegrams
bug ) - the bug that tells you you have 1 new telex after just reading it is
not fixed - Scott Ferguson

*version 2.0 beta 1 release 6*
Three letter abbreviations for all ships in ship.c - Chad Zabel
the following by Kevin Klemmick:
/lib/commands/rea.c                     // fixed telegram's problem reading '%'
/lib/commands/mail.c                    // ditto for mail and wmail
/lib/subs/aircombat.c                   // New air code and bug fixes
/lib/subs/mission.c                     // New air code and bug fixes
/lib/subs/pr.c                          // new prnf function an bug fixes
/lib/subs/paths.c                       // fixes the getpath() bug.
added NEWPAF and HIDDEN to options.h ( commented out for now until
testing is complete ) - Ken H.

*version 2.0 beta 1 release 7*
Fix for fix for exec bug ( lib/player/player.c ) - Chad Zabel
fixed minor bug in aircombat.c ( missing quote on line 613 ) - Ken H.

*version 2.0 beta 1 release 8*
lib/commands/mfir.c Bugfix... in quiet_bigdef, added in a check
for forts >60%.  They were getting cheated out of their bonus in the return
fire process. - Chad Zabel
AIX Port: files
lib/lwp/{GNUmakefile,arch.c,lwp.c,lwpint.h} and                             
make/Systems/aix modified
 - All changes are surrounded by #ifdef AIX32. 
 - one change to lwp.c that will apply to everyone one error
   message changed so it says underflow instead of overflow in the stack
   checking code.
- Bill Canning

*version 2.0 beta 1 release 9*
You can now specify how many mil to board a ship with (wishlist item a1) so
you don't strand your ship.  This is implemented as another parameter
in the board command.  Documentation for board updated too!  - Chad Zabel

- Addition of the HIDDEN and NEWPAF Mods (Major addition). - Kevin Klemmick

*version 2.0 beta 1 release 10*
lib/player/player.c compilation under linux was broken by previous patches
this is fixed now - Ken H.

*version 2.0 beta 1 release 11*
fixed compilation under sunos (lib/lwp/lwp.c had a couple odd '\' characters
in it that caused the sun compiler to complain) - Ken H.

*version 2.0 beta 1 release 12*
added 3 character abbreviations for land unit names - Chad Zabel
fixed bug in "show sect build" command - Ken Huisman
added check in neweff command to see if sector is stopped - Chad Zabel
added test for war declaration before abm launch - Chad Zabel

*version 2.0 beta 1 release 13*
added thread debugging stuff - Dave Pare
changes to autonav code - Chad Zabel
fixes for minor bug in land unit names - Chad Zabel
fix for build.t in docs - Chad Zabel
simple fix for spy command - Chad Zabel
fix for telegrams - Julian Onions
fixup of makefile in util - Julian Onions

*version 2.0 beta 1 release 14*
fix bugs in lwp - Julian Onions
fix bug in edit that caused server crash - Julian Onions
fix bugs in attack command that allowed free resource creation - Chad Zabel
fix bug that made countries in sanctuary lose money - Chad Zabel

*version 2.0 beta 1 release 15*
io bug fixes and misc code changes - Dave Pare
removal of sanctuary print statements to reduce logfile sizes - Ken Huisman
DEC Alpha port - Dave Pare and Ken Stevens

*version 2.0 beta 1 release 16*
made compiling/configuring easier (see READ_THIS_FIRST) - Ken Stevens
fix bug in info.c that caused crash - Janjaap van Velthooven
fix for land units - Chad Zabel

*version 2.0 beta 1 release 17*
major improvements/cleanup of info pages and related files - Ken Stevens
changes to compile scripts - Ken Stevens

*version 2.0 beta 1 release 18*
fixed missile sub abbreviation in ship file - Ken Stevens
added "are you sure" in build command to prevent accidents - Ken Stevens
fixed production simulation report showing "e" sectors producing - Ken Stevens
fixed LandUnits subject not being listed in "info". - Ken Stevens
fixed so unload only works in your sector or ally's sector - Ken Stevens
fixed error in sweep when sector efficiency is printed - Ken Stevens
fixed telegrams reporting bombing damage - Ken Stevens
fixed cede so you can only cede to allies - Ken Stevens
fixed coastwatch to show ship names - Ken Stevens
build command gives informative error message if you are building
in the wrong type of sector or the sector is < 60% eff. - Ken Stevens
build command does not prompt for tech level if not given on command

src/util/land.c
/*
 *  Empire - A multi-player, client/server Internet based war game.
 *  Copyright (C) 1986-2000, Dave Pare, Jeff Bailey, Thomas Ruschak,
 *                           Ken Stevens, Steve McClure
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 *  ---
 *
 *  See the "LEGAL", "LICENSE", "CREDITS" and "README" files for all the
 *  related information and legal notices. It is expected that any future
 *  projects/authors will amend these files as needed.
 *
 *  ---
 *
 *  land.c: New version of empcre - Create the land masses in the game
 * 
 *  Known contributors to this file:
 *     James Anderson, 1987
 *     Dave Pare, 1989
 */

#include <stdio.h>
#include <fcntl.h>
#include "var.h"
#include "misc.h"
#include "power.h"
#include "nat.h"
#include "sect.h"
#include "gamesdef.h"
#include "file.h"
#include "xy.h"
#include "prototypes.h"

#define rnd(x) (random() % (x))

#define XPLATES		((WORLD_X) / 2)	/* basically world x-y size */
#define YPLATES		(WORLD_Y)
#define	XSIZE		(XPLATES)
#define YSIZE		(YPLATES)

#define BIGV		256	/* used in making altitude */
#define SMALLV		128	/* ex-ocean: rnd(SMALLV) - rnd(BIGV) */

typedef enum plates {
    OCEAN, ISLAND, CONTINENT
} plate_e;

#define LANDMIN		1	/* plate altitude for normal land */
#define HILLMIN		34	/* plate altitude for hills */
#define PLATMIN		36	/* plate altitude for plateau */
#define HIGHMIN		98	/* plate altitude for mountains */

#define LANDCH		60	/* land plate percentage */
#define NUMLAND		(YPLATES * XPLATES * LANDCH)/100
#define NUMISLE		NUMLAND/5	/* 1 isle for 5 land */
#define NUMWATER	(XPLATES * YPLATES) - (NUMLAND + NUMISLE)

#define SECTRANGE	3	/* smoothing area */
#define MINCONTDIST	2	/* minimum continent distance */
#define CHUNKSIZE	2	/* basic land block size */
#define NEWCONTDIST	(rnd(mincontdist) + mincontdist)
					/* dist away from others for newcont */
#define NUMCHUNKS	70	/* number of CHUNKS per cont */

struct sctstr sects[YSIZE][XSIZE];

plate_e plates[YPLATES][XPLATES];
int world[YSIZE][XSIZE];

int sectrange = SECTRANGE;
int mincontdist = MINCONTDIST;
int chunksize = CHUNKSIZE;
int numchunks = NUMCHUNKS;
int numisle = NUMISLE;

static void initworld();
static void growcont();
static void newcont();
static int verify();
static int makeland();
static void make_altitude();
static int total_land();
static void make_sects();
static void land_sct_init(coord x, coord y, s_char *ptr);


int
main(argc, argv)
int argc;
s_char **argv;
{
    extern struct empfile empfile[];
    register int n;
    int x, y;
    int i, j;
    time_t now;
    int fd;
    int left;
    int big;

    if (argc > 1 && argc != 6) {
	printf("usage: %s sectrange mincontdist chunksize numchunks numisle\n",
	       *argv);
	return -1;
    }
    if (argc == 6) {
	sectrange = atoi(argv[1]);
	mincontdist = atoi(argv[2]);
	chunksize = atoi(argv[3]);
	numchunks = atoi(argv[4]);
	numisle = atoi(argv[5]);
    }
    printf("sectrange: %d\n", sectrange);
    printf("mincontdist: %d\n", mincontdist);
    printf("chunksize: %d\n", chunksize);
    printf("numchunks: %d\n", numchunks);
    printf("numisle: %d\n", numisle);
    fd = open(empfile[EF_SECTOR].file, O_RDWR | O_CREAT | O_TRUNC, 0660);
    if (fd < 0) {
	perror(empfile[EF_SECTOR].file);
	return -1;
    }
    time(&now);
    srandom(now + getpid());
    initworld((plate_e *)plates);
    left = NUMLAND;
    printf("Creating continents");
    while (left > 0) {
	big = (left / (numchunks * 3 * chunksize * chunksize)) + 1;
	for (n = 0; n < big; n++) {
	    newcont(plates, &x, &y, NEWCONTDIST);
	    left -= makeland(plates, x, y, chunksize, CONTINENT);
	}
	for (n = 0; n < big * numchunks; n++) {
	    growcont(plates, &x, &y);
	    left -= makeland(plates, x, y, chunksize, CONTINENT);
	    if ((n % numchunks) == 0) {
		printf(".");
		fflush(stdout);
	    }
	}
    }
    printf("\n");
    printf("Creating islands");
    for (n = 0; n < numisle; n++) {
	/* find an open spot */
	while (plates[(i = rnd(YPLATES))][(j = rnd(XPLATES))] != OCEAN) ;
	plates[i][j] = ISLAND;
    }
    printf("\n");
    printf("Making altitude\n");
    make_altitude(plates, world);
    printf("Creating sectors\n");
    make_sects(world, sects);
    printf("Writing sectors\n");
    n = write(fd, sects, sizeof(sects));
    if (n < 0) {
	perror(empfile[EF_SECTOR].file);
	return -1;
    }
    if (n != sizeof(sects)) {
	printf("%s: partial write\n", empfile[EF_SECTOR].file);
	return -1;
    }
    close(fd);
    exit(0);
}

static void
initworld(plates)
register plate_e *plates;
{
    register int i;

    for (i = 0; i < XPLATES * YPLATES; i++)
	*plates++ = OCEAN;
}

static void
growcont(plates, xp, yp)
register plate_e plates[YPLATES][XPLATES];
int *xp;
int *yp;
{
    int x, y;

    /* merge with another one */
    while (plates[(y = rnd(YPLATES))][(x = rnd(XPLATES))] == OCEAN) ;
    *xp = x;
    *yp = y;
}

static void
newcont(plates, xp, yp, dist)
register plate_e plates[YPLATES][XPLATES];
int *xp;
int *yp;
int dist;
{
    register int x, y;
    int i;

    for (i = 0; i < 30; i++) {
	y = rnd(YPLATES);
	x = rnd(XPLATES);
	if (verify(plates, x, y, OCEAN, dist))
	    break;
    }
    if (i == 30) {
	growcont(plates, xp, yp);
    } else {
	*xp = x;
	*yp = y;
    }
}

/*
 * verify that with "dist", there are only type "what" sectors
 * returns 0 if fail, 1 if success.
 */
static int
verify(plates, x, y, what, dist)
register plate_e plates[YPLATES][XPLATES];
int x;
int y;
int what;
int dist;
{
    register int xbase, ybase;
    register int x1, y1;

    for (ybase = y - dist; ybase <= y + dist; ybase++) {
	for (xbase = x - dist; xbase <= x + dist; xbase++) {
	    /* normalize to world coords */
	    y1 = ybase < 0 ? ybase + YPLATES : ybase % YPLATES;
	    x1 = xbase < 0 ? xbase + XPLATES : xbase % XPLATES;
	    if (plates[y1][x1] != what)
		return 0;
	}
    }
    return 1;
}

static int
makeland(plates, x, y, dist, what)
register plate_e plates[YPLATES][XPLATES];
int x;
int y;
int dist;
int what;
{
    register int xbase, ybase;
    register int xfail, yfail;
    register int x1, y1;
    int created;

    created = 0;
    for (ybase = y - dist; ybase <= y + dist; ybase++) {
	yfail = y - ybase;
	if (yfail < 0)
	    yfail = -yfail;
	y1 = ybase < 0 ? ybase + YPLATES : ybase % YPLATES;
	for (xbase = x - dist; xbase <= x + dist; xbase++) {
	    x1 = xbase < 0 ? xbase + XPLATES : xbase % XPLATES;
	    if (plates[y1][x1] != OCEAN)
		continue;
	    xfail = x - x1;
	    if (xfail < 0)
		xfail = -xfail;
	    if (xfail < yfail)
		xfail = yfail;
	    if (xfail < dist - 1 || !rnd(xfail + 1) || !rnd(xfail + 1)) {
		plates[y1][x1] = what;
		created++;
	    }
	}
    }
    return created;
}

static void
make_altitude(plates, world)
register plate_e plates[YPLATES][XPLATES];
register int world[YSIZE][XSIZE];
{
    register int x, y;

    for (y = 0; y < YPLATES; y++) {
	for (x = 0; x < XPLATES; x++) {
	    switch (plates[y][x]) {
	    case OCEAN:
				/*-BIGV, -SMALLV/2, SMALLV*/
		world[y][x] = rnd(SMALLV) - rnd(BIGV);
		break;
	    case ISLAND:
				/*-BIGV, 0, BIGV*/
		world[y][x] = rnd(BIGV) - rnd(BIGV) + 2;
		break;
	    case CONTINENT:
				/*-SMALLV, SMALLV/2, BIGV*/
		world[y][x] = rnd(BIGV) - rnd(SMALLV);
	    }
	}
    }
}

static int
total_land(world, xbase, ybase, range)
register int world[YSIZE][XSIZE];
register int xbase;
int ybase;
register int range;
{
    register int x;
    register int xmax;
    register int total;
    register int *row;
    int y;
    int ymax;

    total = 0;
    xmax = xbase + range;
    ymax = ybase + range;
    for (y = ybase; y < ymax; y++) {
	row = world[y % YSIZE];
	for (x = xbase; x < xmax; x++)
	    total += row[x % XSIZE];
    }
    return total;
}

static void
make_sects(world, sects)
register int world[YSIZE][XSIZE];
struct sctstr *sects;
{
    register struct sctstr *sct;
    register int i;
    register int x, y;
    int elev[12 + 12 + 3];	/* # sects from -12 to 12 in steps of 10 elev */
    int range;
    int rangesq;
    int total;
    int sum;

    for (i = 0; i < 12 + 12 + 3; i++)
	elev[i] = 0;
    sum = 0;
    sct = sects;
    for (y = 0; y < YSIZE; y++) {
	for (x = 0; x < XSIZE; x++, sct++) {
	    land_sct_init(x * 2 + (y & 01), y, (s_char *)sct);
	    range = 3 + rnd(sectrange);
	    rangesq = range * range;
	    total = total_land(world, x, y, range) / rangesq;
	    if (total < LANDMIN) {
		sct->sct_type = SCT_WATER;
	    } else if (total < HILLMIN)
		sct->sct_type = SCT_RURAL;
	    else if (total < PLATMIN)
		sct->sct_type = SCT_MOUNT;
	    else if (total < HIGHMIN)
		sct->sct_type = SCT_RURAL;
	    else
		sct->sct_type = SCT_MOUNT;
	    sct->sct_elev = total;
	    sct->sct_newtype = sct->sct_type;
	    sum += total;
	    if (total < -129)
		elev[0]++;
	    else if (total > 129)
		elev[26]++;
	    else
		elev[13 + total / 10]++;
	}
    }
    for (i = 0; i < 12 + 12 + 3; i++)
	if (elev[i] != 0)
	    printf("%4d sectors elevation %4d to %4d\n",
		   elev[i], 10 * i - 140, 10 * i - 130);
}

static void
land_sct_init(coord x, coord y, s_char *ptr)
{
    struct sctstr *sp = (struct sctstr *)ptr;

    sp->ef_type = EF_SECTOR;
    sp->sct_x = x;
    sp->sct_y = y;
    sp->sct_dist_x = x;
    sp->sct_dist_y = y;
}
